File: src/main/java/com/starshootercity/abilities/AquaAffinity.java
package com.starshootercity.abilities;

import com.starshootercity.OriginSwapper;
import net.kyori.adventure.key.Key;
import org.bukkit.entity.Player;
import org.bukkit.potion.PotionEffectType;
import org.jetbrains.annotations.NotNull;

import java.util.List;

public class AquaAffinity implements VisibleAbility, BreakSpeedModifierAbility {
    @Override
    public @NotNull Key getKey() {
        return Key.key("origins:aqua_affinity");
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getDescription() {
        return OriginSwapper.LineData.makeLineFor("You may break blocks underwater as others do on land.", OriginSwapper.LineData.LineComponent.LineType.DESCRIPTION);
    }

    @Override
    public @NotNull List<OriginSwapper.LineData.LineComponent> getTitle() {
        return OriginSwapper.LineData.makeLineFor("Aqua Affinity", OriginSwapper.LineData.LineComponent.LineType.TITLE);
    }

    @Override
    public BlockMiningContext provideContextFor(Player player) {
        return new BlockMiningContext(
                player.getInventory().getItemInMainHand(),
                player.getPotionEffect(PotionEffectType.SLOW_DIGGING),
                player.getPotionEffect(PotionEffectType.FAST_DIGGING),
                player.getPotionEffect(PotionEffectType.CONDUIT_POWER),
                true,
                true,
                true
        );
    }

    @Override
    public boolean shouldActivate(Player player) {
        return player.isInWater();
    }
}

File: core/src/main/kotlin/ru/turbovadim/abilities/AquaAffinity.kt
package ru.turbovadim.abilities

import ru.turbovadim.OriginSwapper.LineData.Companion.makeLineFor
import ru.turbovadim.OriginSwapper.LineData.LineComponent
import ru.turbovadim.OriginsRebornEnhanced.Companion.NMSInvoker
import ru.turbovadim.abilities.BreakSpeedModifierAbility.BlockMiningContext
import net.kyori.adventure.key.Key
import org.bukkit.entity.Player
import org.bukkit.potion.PotionEffectType

class AquaAffinity : VisibleAbility, BreakSpeedModifierAbility {
    override fun getKey(): Key {
        return Key.key("origins:aqua_affinity")
    }

    override val description: MutableList<LineComponent> = makeLineFor("You may break blocks underwater as others do on land.", LineComponent.LineType.DESCRIPTION)

    override val title: MutableList<LineComponent> = makeLineFor("Aqua Affinity", LineComponent.LineType.TITLE)

    override fun provideContextFor(player: Player) = with(player) {
        BlockMiningContext(
            inventory.itemInMainHand,
            getPotionEffect(NMSInvoker.hasteEffect),
            getPotionEffect(NMSInvoker.miningFatigueEffect),
            getPotionEffect(PotionEffectType.CONDUIT_POWER),
            true,
            true,
            true
        )
    }

    override fun shouldActivate(player: Player): Boolean {
        return player.isInWater
    }
}
